### PR TITLE
Petsc 3.18 fixes

### PR DESCRIPTION
**Pull request by roystgnr**
This along with the next libMesh submodule update will take care of #22549.  We might wait until that's ready and merge this branch into the submodule update PR branch, but I'd like to make sure in advance that the changes here don't have any side effects that CI can catch.

## Reason
Compatibility with PETSc 3.18 and later

## Design
Fixes for bad code that PETSc 3.18 now warns about; shims and `#ifdef` clauses for perfectly fine code that PETSc just doesn't support anymore.

## Impact
There should be no external impact, we'll just be able to compile against new PETSc releases again.

### CODE REVIEW

**Comment by loganharbour** [comment]
We don't officially support anything older than PETSc 3.11.4, and have made the decision in the past not to test anything earlier than that. Who is the customer for this?

File: framework/src/utils/PetscDMMoose.C [hunk 1]
```diff
@@ -37,6 +37,15 @@
 #include "libmesh/preconditioner.h"
 #include "libmesh/elem_side_builder.h"
 
+// As of 3.18, %D is no longer supported in format strings, but the
+// replacement PetscInt_FMT didn't get added until 3.7.2, and the
+// libMesh shim hasn't hit our submodule yet
+#if PETSC_RELEASE_LESS_THAN(3, 8, 0)
```
Comment (lindsayad) [inline framework/src/utils/PetscDMMoose.C:43]: Is this 3.18 or 3.8?

**Review by lindsayad — CHANGES_REQUESTED** [review]

**Comment by roystgnr** [comment]
Yes, of course I mean 3.18.  Damn it.

**Comment by loganharbour** [comment]
> Yes, of course I mean 3.18. Damn it.

60a4fa00-41db-11eb-90c2-e18f174d8468

**Review by loganharbour — COMMENTED** [review]
We should probably do the libmesh + petsc 3.18 update all in one go here, no?

**Comment by roystgnr** [comment]
Ah, damn it, I get it.  I was copying fixes from the libMesh patches.  In the libMesh format string shim I switched to PetscInt_FMT at PETSc 3.8+ because that's when they added that macro.  But then in MOOSE I also *copied* that shim to other updates that really don't become possible until 3.18.

**Comment by roystgnr** [comment]
> We should probably do the libmesh + petsc 3.18 update all in one go here, no?

That's what I was leaning towards in the description, yeah.  But the next libMesh update has other things it's waiting on, so I'm not pulling this PR yet.  I think my "I'd like to make sure in advance that the changes here don't have any side effects that CI can catch." strategy has been sadly vindicated...

**Comment by lindsayad** [comment]
It would be great to do a PETSc update. But I will say that can sometimes turn into quite a task. @roystgnr do you want to take on that responsibility? Otherwise I can. I think it might be a good idea to try and segregate the petsc and libMesh updates if possible so we can more easily discern possible test failure causes

**Comment by roystgnr** [comment]
Ok, I parsed "petsc 3.18 update" as "petsc 3.18 compatibility update", and I'm now guessing that isn't what was intended?

I don't actually know if there's any reason to update our PETSc modules to 3.18.  Neither "Let's have a module to cover compatibility with every possible PETSc version" nor "Let's not bother with compatibility with PETSc versions we don't have in CI" seems like a good idea to me.  I do want to try to support end users who grab the latest PETSc, but they make API-breaking changes with arbitrary minor version increments, and I don't want to try to bother maintaining software stacks and test recipes to keep up with the history of that.

**Comment by lindsayad** [comment]
What about updating our module hash to the last 3.17 minor version?

**Comment by lindsayad** [comment]
(in a different PR by a yet-decided author)

**Comment by roystgnr** [comment]
>What about updating our module hash to the last 3.17 minor version?

I'm not opposed to it (or to updating to 3.18 if someone wants to), but what do we gain from it?  I see clear reasons (albeit not quite as urgent on the MOOSE side now that so many users are on our stack) to update the list of *supported* PETSc versions as fast as they come out, but I'm not sure what we get from making an update on our *supplied* PETSc versions except when there's a new feature that somebody wants.

**Comment by garciapintado** [comment]
Well, at least one user (me) will immediately benefit from having PETSc 3.18 compatibility. See #22565

**Comment by lindsayad** [comment]
@roystgnr is drawing a distinction between compatibility and what we provide in our submodule. He is adding the compatibility here which should be immediately beneficial to you @garciapintado. The latter point should not really be relevant to you since you already have a built PETSc 3.18

As far as what we provide ... I suppose it is not all that important until a user wants a particular feature that our supplied petsc doesn't provide. The fact that @garciapintado wants to use PETSc 3.18 could mean that such a feature already exists. I would say that most of our users use our provided PETSc, whether that comes from building it from the submodule hash or through conda which uses the submodule hash. In that vein providing a new PETSc that has all the best features (and maybe some bug fixes from previous major versions) sounds like a good thing to me.

**Comment by loganharbour** [comment]
I think the issue is that with our infrastructure it's easier to test new PETSc by making it the default. Which probably isn't the best scenario... but it is the case

**Comment by roystgnr** [comment]
>He is adding the compatibility here which should be immediately beneficial to you @garciapintado.

(wincing) ... for some definition of "immediately".  You'll also need to use a more recent libMesh than the current MOOSE submodule.

**Comment by garciapintado** [comment]
> PETSc

No, I have not a built PETSc 3.18. The issue referred to by discussion #22565 is that I have tried to build MOOSE with its default PETSc built (3.16.6 if I'm right) in a HPC. This PETSc compilation has failed and I've sent the log files to a PETSc support e-mail. A PETSc developer has answered. He has not been very explicit, but indicated that the compilation error would be solved if I use PETSc 3.18. I'd presume that his answer comes from looking at the compilation *.log files I've sent them, and that he must have his reasons to indicate so.  That's all I can say. I have MOOSE (with PETSc 3.16) installed in other personal iMAC and linux systems without any problem. But there must be something in this HPC cluster, which appears to require later PETSc versions, according to the PETSc developer.

**Comment by roystgnr** [comment]
> providing a new PETSc that has all the best features (and maybe some bug fixes from previous major versions) sounds like a good thing to me.

If it were costless, I'd agree.  Is it?  I got the impression from https://github.com/idaholab/moose/pull/22062#issuecomment-1243734954 that frequent module updates in our conda stack aren't free of annoyance.  I'm already pushing libMesh updates every month (and I'd be doing it more frequently if I thought it was costless); PETSc updates also come along around once a month, so if users need to worry about dependency updates we'd be doubling the rate of hassle.

(all that said, it's debate about the future - if our current petsc hash has a bug triggering for users then we need to update that ASAP, though I'd still want to get this in first and get a separate libMesh submodule update second)

> We should probably do the libmesh + petsc 3.18 update all in one go here, no?

And now I'd say no: if we've got users already being bitten by a 3.16 bug then I'd want to merge this ASAP to at least give a possible manual path forward.  I'll try to rush the next libMesh submodule update though.

**Comment by loganharbour** [comment]
>  I'm already pushing libMesh updates every month (and I'd be doing it more frequently if I thought it was costless); PETSc updates also come along around once a month, so if users need to worry about dependency updates we'd be doubling the rate of hassle.

Sounds like the perfect opportunity to update both together.

File: framework/src/utils/PetscDMMoose.C [hunk 2]
```diff
@@ -1496,7 +1508,7 @@ DMView_Moose(DM dm, PetscViewer viewer)
     CHKERRQ(ierr);
     for (const auto & bit : *(dmm->_block_ids))
     {
-      ierr = PetscViewerASCIIPrintf(viewer, "(%s,%D) ", bit.first.c_str(), bit.second);
+      ierr = PetscViewerASCIIPrintf(viewer, "(%s,%d) ", bit.first.c_str(), bit.second);
```
Comment (lindsayad) [inline framework/src/utils/PetscDMMoose.C:1511]: I believe this should be unsigned. I only thought of this because you changed to `u` for the `_var_ids`

Comment (lindsayad) [reply framework/src/utils/PetscDMMoose.C:1511]: Oops never mind I didn't read `id_types.h` close enough. Nothing to see here

File: framework/src/utils/PetscDMMoose.C [hunk 3]
```diff
@@ -1988,7 +1996,10 @@ DMSetUp_Moose(DM dm)
 
 #undef __FUNCT__
 #define __FUNCT__ "DMSetFromOptions_Moose"
-#if !PETSC_VERSION_LESS_THAN(3, 7, 0)
+#if !PETSC_VERSION_LESS_THAN(3, 18, 0)
+PetscErrorCode
+DMSetFromOptions_Moose(DM dm, PetscOptionItems * /*options*/) // >= 3.8.0
```
Comment (lindsayad) [inline framework/src/utils/PetscDMMoose.C:2001]:
```suggestion
DMSetFromOptions_Moose(DM dm, PetscOptionItems * /*options*/) // >= 3.18.0
```

**Review by lindsayad — CHANGES_REQUESTED** [review]

**Comment by lindsayad** [comment]
> I'm already dreading the first time we see an exodiff and don't even know whether it's from a libMesh change or a PETSc change, though.

Right, this is a potential developer nightmare. I suppose we are balancing a nightmare for the developer with incursions on the user

**Review by lindsayad — APPROVED** [review]

**Comment by roystgnr** [comment]
"Failed but allowed" appears to mean "there's an unrelated compile problem in mockingbird", so I'm going to say the changes here are safe enough.

**Comment by cticenhour** [comment]
I am suddenly getting lots of framework build failures in CI due to new compilation warnings stemming from `PetscDMMoose.C`. I am surprised these have not shown themselves in next. See https://civet.inl.gov/pr/16894/ for some examples. Note that it is not happening on every target where framework is being built - haven't looked closely at the compiler modules being loaded for those yet, but it might simply be a difference in compiler being used. Regardless, flagging here since this just went in.

**Comment by roystgnr** [comment]
Oh, that's wonderful.  The previous version of `PetscOptionsBegin` started with `0;`, probably for backwards compatibility with code like our `ierr = PetscOptionsBegin`, and now that such code is broken with new PETSc, removing the `ierr =` is triggering a "statement has no effect" warning with old PETSc.

**Comment by cticenhour** [comment]
OK - are you able to make a patch for this? I am available for a review.

**Comment by roystgnr** [comment]
I'll give it a try.  I don't actually have an older PETSc build around anymore, though, so I'm not sure how to test the patch.  How did we not see any of those failures in the CI for *this* PR?

**Comment by cticenhour** [comment]
I'm not sure about that either - I don't think we've changed any modules on the Linux CI environments in a while.

**Comment by cticenhour** [comment]
If you DM be a branch - I can always cherry pick and place it in the PR I mentioned earlier as a test of it. I've been hitting this consistently today.

**Comment by roystgnr** [comment]
Give https://github.com/roystgnr/moose/tree/petsc_38_fix_fixes a crack?

**Comment by cticenhour** [comment]
Will do - it looks as though this PR used an available module of PETSc for targets like https://civet.inl.gov/job/1222811/ where my PR opts for the submodule being built instead (as I am making conda package changes). So I think that logic between the conda package state and the pre-built module is tied into this somehow.

**Comment by cticenhour** [comment]
A previous failure is now passing with that change: https://civet.inl.gov/job/1224154/. Thanks @roystgnr!

**Pull request merged**